File: modules/core/src/main/java/org/torquebox/clojure/core/as/ClojureRuntimeInstaller.java
/*
 * Copyright 2008-2011 Red Hat, Inc, and individual contributors.
 * 
 * This is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 * 
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this software; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
 */

package org.torquebox.clojure.core.as;

import org.jboss.as.server.deployment.Attachments;
import org.jboss.as.server.deployment.DeploymentPhaseContext;
import org.jboss.as.server.deployment.DeploymentUnit;
import org.jboss.as.server.deployment.DeploymentUnitProcessingException;
import org.jboss.as.server.deployment.DeploymentUnitProcessor;
import org.jboss.logging.Logger;
import org.jboss.modules.Module;
import org.torquebox.clojure.core.ClojureApplicationMetaData;
import org.torquebox.clojure.core.ClojureRuntime;

/**
 * Handle mounting .clj files and marking them as a DEPLOYMENT_ROOT
 * FIXME: This doesn't handle archives
 * 
 */
public class ClojureRuntimeInstaller implements DeploymentUnitProcessor {
    
    public ClojureRuntimeInstaller() {
    }

    @Override
    public void deploy(DeploymentPhaseContext phaseContext) throws DeploymentUnitProcessingException {
        DeploymentUnit deploymentUnit = phaseContext.getDeploymentUnit();

        ClojureApplicationMetaData metaData = deploymentUnit.getAttachment( ClojureApplicationMetaData.ATTACHMENT_KEY );
        
        if (metaData == null) {
            return;
        }
        
        Module module = deploymentUnit.getAttachment( Attachments.MODULE );
        
        deploymentUnit.putAttachment( ClojureRuntime.ATTACHMENT_KEY, new ClojureRuntime( module.getClassLoader() ) );
    }

    @Override
    public void undeploy(DeploymentUnit context) {
       
    }
        
    static final Logger log = Logger.getLogger( "org.torquebox.clojure.core.as" );
}

File: modules/core/src/test/clojure/org/fnbox/core/processors/test/ClojureRuntimeInstaller.clj
(ns org.fnbox.core.processors.test.ClojureRuntimeInstaller
  (:use clojure.test)
  (:use fnbox.test.as.helpers)
  (:import [org.jboss.as.server.deployment Attachments])
  (:import [org.jboss.modules Module])
  (:import [org.fnbox.core.processors ClojureRuntimeInstaller])
  (:import [org.fnbox.core ClojureMetaData ClojureRuntime])
  (:require [clojure.java.io :as io]))

(use-fixtures :each
              (harness-with [(ClojureRuntimeInstaller.)]))

(deftest it-should-install-a-runtime-if-metadata-is-present
  (let [phase-context (.createPhaseContext *harness*)
        unit (.getDeploymentUnit phase-context)
        module (Module/getSystemModule)]
    (doto unit
      (.putAttachment ClojureMetaData/ATTACHMENT_KEY (ClojureMetaData. "foo" {}))
      (.putAttachment Attachments/MODULE module))
    
    (.deploy *harness* phase-context)
    
    (let [runtime (.getAttachment unit ClojureRuntime/ATTACHMENT_KEY)]
      (is (not (nil? runtime)))
      (is (= (.getClassLoader module) (.getClassLoader runtime))))))

(deftest it-should-not-install-a-runtime-if-no-metadata-present
  (let [phase-context (.createPhaseContext *harness*)
        unit (.getDeploymentUnit phase-context)]
    (.deploy *harness* phase-context)
    (is (nil? (.getAttachment unit ClojureRuntime/ATTACHMENT_KEY)))))
